File: SignalR Simple Users On Line Counter/[C#][VB]-SignalR Simple Users On Line Counter/C# and VB/C#,VB/SimpleUsersOnLineVB/App_Code/UsersOnLine.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports Microsoft.AspNet.SignalR

Namespace SimpleUsersOnLine
    Public Class UsersOnLine
        Inherits Hub
        Public Shared users As New List(Of String)()

        Public Overrides Function OnConnected() As System.Threading.Tasks.Task
            Dim clientId As String = GetClientId()

            If users.IndexOf(clientId) = -1 Then
                users.Add(clientId)
            End If

            ShowUsersOnLine()

            Return MyBase.OnConnected()
        End Function

        Public Overrides Function OnReconnected() As System.Threading.Tasks.Task
            Dim clientId As String = GetClientId()
            If users.IndexOf(clientId) = -1 Then
                users.Add(clientId)
            End If

            ShowUsersOnLine()

            Return MyBase.OnReconnected()
        End Function

        Public Overrides Function OnDisconnected() As System.Threading.Tasks.Task
            Dim clientId As String = GetClientId()

            If users.IndexOf(clientId) > -1 Then
                users.Remove(clientId)
            End If

            ShowUsersOnLine()

            Return MyBase.OnDisconnected()
        End Function


        Private Function GetClientId() As String
            Dim clientId As String = ""
            If Not (Context.QueryString("clientId") Is Nothing) Then
                'clientId passed from application
                clientId = Context.QueryString("clientId").ToString()
            End If

            If clientId.Trim() = "" Then
                'default clientId: connectionId
                clientId = Context.ConnectionId
            End If
            Return clientId

        End Function
        Public Sub Log(message As String)
            Clients.All.log(message)
        End Sub
        Public Sub ShowUsersOnLine()
            Clients.All.showUsersOnLine(users.Count)
        End Sub
    End Class
End Namespace

File: SignalR Simple Users On Line Counter/[C#][VB]-SignalR Simple Users On Line Counter/C# and VB/C#,VB/SimpleUsersOnLine/UsersOnLine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using Microsoft.AspNet.SignalR;

namespace SimpleUsersOnLine
{
    public class UsersOnLine : Hub
    {
        static List<string> users=new List<string>();

        public override System.Threading.Tasks.Task OnConnected()
        {
            string clientId = GetClientId();
            
            if (users.IndexOf(clientId)==-1) {
                users.Add(clientId);
            }

            ShowUsersOnLine();
            
            return base.OnConnected();
        }

        public override System.Threading.Tasks.Task OnReconnected()
        {
            string clientId = GetClientId();
            if (users.IndexOf(clientId) == -1)
            {
                users.Add(clientId);
            }

            ShowUsersOnLine();
        
            return base.OnReconnected();
        }

        public override System.Threading.Tasks.Task OnDisconnected()
        {
            string clientId = GetClientId();
            
            if (users.IndexOf(clientId) > -1)
            {
                users.Remove(clientId);
            }

            ShowUsersOnLine();

            return base.OnDisconnected();
        }


        private string GetClientId() {
            string clientId = "";
            if (!(Context.QueryString["clientId"] == null))
            {
                //clientId passed from application
                clientId = Context.QueryString["clientId"].ToString();
            }

            if (clientId.Trim() == "")
            {
                //default clientId: connectionId
                clientId = Context.ConnectionId;
            }
            return clientId;
        
        }
        public void Log(string message) {
            Clients.All.log(message);
        }
        public void ShowUsersOnLine()
        {
            Clients.All.showUsersOnLine(users.Count);
        }
    }
}